### PR TITLE
[tt-train] Wire up or disable vocab parallel cross entropy loss across all trainers

### PR DESCRIPTION
**Pull request by bklockiewiczTT**
## Motivation
[#42579](https://github.com/tenstorrent/tt-metal/pull/42579) (`[tt-train] Vocab parallel cross entropy loss`) introduced a vocab-parallel CE op, but none of the Python trainers were actually using it — under TP they still computed loss on the full vocab logits, which is expensive and defeats the point of the op. This PR plumbs the op through the trainers (nano_gpt, lora_llama, SFTTrainer, multihost pipeline/hierarchical) so TP runs pick it up automatically, and adds the `ReduceType::NONE` mode the multihost trainers need.

## Summary
- Wires Python training paths (nano_gpt, lora_llama, multihost pipeline/hierarchical, SFTTrainer) through vocab-parallel cross-entropy when tensor parallelism is active, so loss is computed without materializing the full vocab logits.
- Extends `vocab_parallel_cross_entropy_loss` with a `ReduceType::NONE` mode and exposes it via nanobind; SFTTrainer auto-selects it under TP.
- Adds a C++ test (`vocab_parallel_cross_entropy_loss_test.cpp`) and Python test (`test_vocab_parallel_loss.py`) covering the new loss paths.
- Misc: fixes `tp_axis` in nano_gpt, small clarifying comment in the GRPO Llama completer.

## Notes for reviewers
- Follow the commits

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:bklockiewicz/voc-par-ce-pyth)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:bklockiewicz/voc-par-ce-pyth)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:bklockiewicz/voc-par-ce-pyth)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:bklockiewicz/voc-par-ce-pyth)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:bklockiewicz/voc-par-ce-pyth)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:bklockiewicz/voc-par-ce-pyth)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:bklockiewicz/voc-par-ce-pyth)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:bklockiewicz/voc-par-ce-pyth)

### CODE REVIEW

File: tt-train/sources/ttml/ttml/common/model_factory.py [hunk 1]
```diff
@@ -168,3 +168,18 @@ def create_model(self):
...
+        Mirrors the C++ trainer in nano_gpt/main.cpp: the TP-only paths build
+        ``models.distributed.{llama,gpt2}`` whose LM heads are
+        ``ColumnParallelLinear(gather_output=False)``, so they emit vocab-sharded
+        logits ([B,1,S,padded_V/tp_size] per device) and must be paired with
+        ``ttml.ops.distributed.vocab_parallel_cross_entropy_loss``.  The
+        pipeline-parallel Llama path still hardcodes ``gather_output=true`` at
+        the LM head, so the last stage gets full-vocab logits and plain
+        ``cross_entropy_loss`` is the right call there.
+        """
+        return self.device_config.enable_tp and self.multihost_config.pipeline_parallel_config is None
```
Comment (dyurshevichTT) [inline tt-train/sources/ttml/ttml/common/model_factory.py]: Why not to use sharded loss if pp is enabled? Since you never gather logits it will crash if someone tries to run with pp and tp enabled. When we start working on pp, this check will be annoying to debug, I think its better to always use sharded loss if tp is enabled.

**Review by dyurshevichTT — APPROVED** [review]
Looking good to me.

File: tt-train/sources/ttml/ttml/common/model_factory.py (hunk 1 above)
Comment (bklockiewiczTT) [reply tt-train/sources/ttml/ttml/common/model_factory.py]: Well are you sure it will crash and not silently produce error? My thinking was it's best to disable it with a clear place to turn it on. But what about enabling tp but including an assert or warning?

Comment (philei-tt) [reply tt-train/sources/ttml/ttml/common/model_factory.py]: But why disable for PP though? It should just work, no?

File: tt-train/sources/ttml/models/distributed/pipeline_parallel_llama.cpp [hunk 2]
```diff
@@ -139,8 +139,12 @@ PipelineParallelLlama::PipelineParallelLlama(
     if (is_last_rank()) {
         ln_fc = std::make_shared<ttml::modules::RMSNormLayer>(embedding_dim);
         if (is_tensor_parallel) {
+            fmt::print(
+                "WARNING: Building pipeline-parallel + tensor-parallel LM head with "
+                "gather_output=false. PP+TP end-to-end correctness has not been validated; "
+                "please sanity-check loss values before relying on this path.\n");
```
Comment (philei-tt) [inline tt-train/sources/ttml/models/distributed/pipeline_parallel_llama.cpp:145]: Any specific reason why you're so concerned with this path?

**Review by philei-tt — APPROVED** [review]
Thanks!

File: tt-train/sources/ttml/models/distributed/pipeline_parallel_llama.cpp (hunk 2 above)
Comment (bklockiewiczTT) [reply tt-train/sources/ttml/models/distributed/pipeline_parallel_llama.cpp:145]: No, just out of an abundance of caution.

File: tt-train/sources/ttml/ttml/common/model_factory.py (hunk 1 above)
Comment (dyurshevichTT) [reply tt-train/sources/ttml/ttml/common/model_factory.py]: Yes I think so
